### PR TITLE
Fix bug where functions with type hints caused errors

### LINKED ISSUE #3514: "issubclass() arg 1 must be a class" error appears when using input parameters with Optional type hint
### Describe the bug

I'm getting an error `issubclass() arg 1 must be a class` when using Optional typing in the target function. The problem started appearing with version `3.21.0`

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Reproduction

Steps to reproduce:
1. Install gradio `3.21.0`
2. Target function with the following type hints:
```
def transcribe(
        language: Optional[str] = 'auto',
):
```
3. Run gradio application

AR:
```
  File ".../app.py", line 67, in <module>
    allow_flagging='never',
  File ".../.venv/lib/python3.7/site-packages/gradio/interface.py", line 463, in __init__
    self.attach_submit_events(submit_btn, stop_btn)
  File ".../.venv/lib/python3.7/site-packages/gradio/interface.py", line 624, in attach_submit_events
    postprocess=not (self.api_mode),
  File ".../.venv/lib/python3.7/site-packages/gradio/events.py", line 147, in __call__
    trigger_only_on_success=self.trigger_only_on_success,
  File ".../.venv/lib/python3.7/site-packages/gradio/blocks.py", line 190, in set_event_trigger
    check_function_inputs_match(fn, inputs, inputs_as_dict)
  File ".../.venv/lib/python3.7/site-packages/gradio/utils.py", line 943, in check_function_inputs_match
    if not is_special_typed_parameter(name):
  File ".../.venv/lib/python3.7/site-packages/gradio/utils.py", line 932, in is_special_typed_parameter
    is_event_data = issubclass(parameter_types.get(name, int), EventData)
TypeError: issubclass() arg 1 must be a class
```

The problem is that the first parameter for `issubclass` is `typing.Union[str, NoneType]`

### Screenshot

_No response_

### Logs

_Stacktrace is attached in the reproduction steps._

### System Info

```shell
Gradio 3.21.0, Python 3.7
```

### Severity

serious, but I can work around it by removing annotations

### PR DESCRIPTION
**Pull request by freddyaboulton**
# Description

Fixes #3514

End to end test
```python
import gradio as gr
from typing import Literal

def calculator(num1,
               operation: Literal["add", "subtract", "multiply", "divide"],
               num2):
    if operation == "add":
        return num1 + num2
    elif operation == "subtract":
        return num1 - num2
    elif operation == "multiply":
        return num1 * num2
    elif operation == "divide":
        if num2 == 0:
            raise gr.Error("Cannot divide by zero!")
        return num1 / num2

demo = gr.Interface(
    calculator,
    [
        "number", 
        gr.Radio(["add", "subtract", "multiply", "divide"]),
        "number"
    ],
    "number",
    examples=[
        [5, "add", 3],
        [4, "divide", 2],
        [-4, "multiply", 2.5],
        [0, "subtract", 1.2],
    ],
    title="Toy Calculator",
    description="Here's a sample toy calculator. Allows you to calculate things like $2+2=4$",
)
if __name__ == "__main__":
    demo.launch()
```

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have added a short summary of my change to the CHANGELOG.md
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# A note about the CHANGELOG

Hello 👋 and thank you for contributing to Gradio!

All pull requests must update the change log located in CHANGELOG.md, unless the pull request is labeled with the "no-changelog-update" label.

Please add a brief summary of the change to the Upcoming Release > Full Changelog section of the CHANGELOG.md file and include
a link to the PR (formatted in markdown) and a link to your github profile (if you like). For example, "* Added a cool new feature by `[@myusername](link-to-your-github-profile)` in `[PR 11111](https://github.com/gradio-app/gradio/pull/11111)`".

If you would like to elaborate on your change further, feel free to include a longer explanation in the other sections.
If you would like an image/gif/video showcasing your feature, it may be best to edit the CHANGELOG file using the 
GitHub web UI since that lets you upload files directly via drag-and-drop.

### CODE REVIEW

File: test/test_utils.py [hunk 1]
```diff
@@ -633,6 +635,16 @@ class GenericObject:
...
 
+    def test_is_special_typed_parameter(self):
+        def func(a: List[str], b: Literal["a", "b"], c, d: Request):
+            pass
+
+        hints = get_type_hints(func)
+        assert not is_special_typed_parameter("a", hints)
+        assert not is_special_typed_parameter("a", hints)
+        assert not is_special_typed_parameter("a", hints)
+        assert not is_special_typed_parameter("a", hints)
```
Comment (abidlabs) [inline test/test_utils.py:646]: I believe this was intended? 
```suggestion
        assert not is_special_typed_parameter("a", hints)
        assert not is_special_typed_parameter("b", hints)
        assert not is_special_typed_parameter("c", hints)
        assert is_special_typed_parameter("d", hints)
```

File: gradio/utils.py [hunk 2]
```diff
@@ -845,22 +845,24 @@ def get_type_hints(fn):
...
+def is_special_typed_parameter(name, parameter_types):
+    from gradio.helpers import EventData
+    from gradio.routes import Request
+
+    """Checks if parameter has a type hint designating it as a gr.Request or gr.EventData"""
+    is_request = parameter_types.get(name, "") == Request
+    # use int in the fall-back as that will always be false
+    hint = parameter_types.get(name, int)
+    is_event_data = inspect.isclass(hint) and issubclass(hint, EventData)
+    return is_request or is_event_data
```
Comment (abidlabs) [inline gradio/utils.py:857]: I know this precedes this PR, but I feel like this is easier to understand:
```suggestion
    hint = parameter_types.get(name)
    if not hint:
        return False
    is_request = hint == Request
    is_event_data = inspect.isclass(hint) and issubclass(hint, EventData)
    return is_request or is_event_data
```

File: test/test_utils.py (hunk 1 above)
Comment (freddyaboulton) [reply test/test_utils.py:646]: Lol yes 🙈

File: gradio/utils.py (hunk 2 above)
Comment (freddyaboulton) [reply gradio/utils.py:857]: Agreed!

**Comment by abidlabs** [comment]
Thanks for addressing the feedback @freddyaboulton! Reconciled with `main` and gonna merge this in after CI passes if you don't mind

**Pull request merged**